src/library/serato/seratofeature.cpp

```diff
@@ -0,0 +1,1043 @@
+    BeatgridLocked = 0x6262676c, // bbgl
+    FileTime = 0x75746d65,       // utme
+    Missing = 0x626d6973,        // bmis
+    Sorting = 0x7472736f,        // osrt
+    ReverseOrder = 0x62726576,   // brev
+    ColumnTitle = 0x6f766374,    // ovct
+    ColumnName = 0x7476636e,     // tvcn
+    ColumnWidth = 0x74766377,    // tvcw
+    TrackPath = 0x7074726b,      // ptrk
+};
+
+struct serato_track_t {
+    QString filetype;
+    QString location;
+    QString title;
+    QString artist;
+    QString album;
+    QString genre;
+    QString comment;
+    QString grouping;
+    QString label;
+    int year = -1;
+    int duration = -1;
+    QString bitrate;
+    QString samplerate;
+    double bpm = -1.0;
+    QString key;
+    bool beatgridlocked = false;
+    bool missing = false;
+    quint32 filetime = 0;
+    quint32 datetimeadded = 0;
+};
+
+const QString kDatabaseDirectory = "_Serato_";
+const QString kDatabaseFilename = "database V2";
+const QString kCrateDirectory = "Subcrates";
+const QString kCrateFilter = "*.crate";
+const QString kSmartCrateDirectory = "Smart Crates";
+const QString kSmartCrateFilter = "*.scrate";
+
+const QString kSeratoLibraryTable = "serato_library";
+const QString kSeratoPlaylistsTable = "serato_library";
+const QString kSeratoPlaylistTracksTable = "serato_library";
+
+int createPlaylist(QSqlDatabase& database, QString name, QString databasePath) {
+    QSqlQuery query(database);
+    query.prepare(
+            "INSERT INTO serato_playlists (name, serato_db)"
+            "VALUES (:name, :serato_db)");
+    query.bindValue(":name", name);
+    query.bindValue(":serato_db", databasePath);
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query) << "databasePath: " << databasePath;
+        return -1;
+    }
+
+    return query.lastInsertId().toInt();
+}
+
+int insertTrackIntoPlaylist(QSqlDatabase& database, int playlistId, int trackId, int position) {
+    QSqlQuery query(database);
+    query.prepare(
+            "INSERT INTO serato_playlist_tracks (playlist_id, track_id, position) "
+            "VALUES (:playlist_id, :track_id, :position)");
+    query.bindValue(":playlist_id", playlistId);
+    query.bindValue(":track_id", trackId);
+    query.bindValue(":position", position);
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+        return -1;
+    }
+
+    return query.lastInsertId().toInt();
+}
+
+inline QString parseText(const QByteArray& data, const quint32 size) {
+    return QTextCodec::codecForName("UTF-16BE")->toUnicode(data, size);
+}
+
+inline bool parseBoolean(const QByteArray& data) {
+    return data.at(0) != 0;
+}
+
+inline quint32 parseUInt32(const QByteArray& data) {
+#if QT_VERSION >= QT_VERSION_CHECK(5, 12, 0)
+    return qFromBigEndian<quint32>(data);
+#else
+    return qFromBigEndian<quint32>(
+            reinterpret_cast<const uchar*>(data.constData()));
+#endif
+}
+
+inline bool parseTrack(serato_track_t& track, QIODevice& buffer) {
```

Comment:
serato_track_t* pSeratoTrack, const QIODevice& buffer

Reply:
I used pointers for both. `buffer` needs to be non-const to call `read()`.